zebra-state/src/service/non_finalized_state.rs

```diff
@@ -45,6 +47,9 @@ pub struct NonFinalizedState {
     /// callers should migrate to `chain_iter().next()`.
     chain_set: BTreeSet<Arc<Chain>>,
 
+    /// Invalidated blocks and descendants
+    invalidated_blocks: HashMap<block::Hash, InvalidatedBlockData>,
```

Comment:
Optional/Nitpick: Ideally this hash map or its values would be in an `Arc`, because Zebra clones the non-finalized state fairly frequently. It shouldn't be possible to update this field without access to the RPC server, so this would be cleanup, not a security issue.

I'm also wondering if this could be simplified to be `HashMap<Hash, Arc<Vec<ContextuallyValidatedBlock>>>` where the first item in the value is the explicitly invalidated block?

```suggestion
    /// Blocks that have been invalidated in, and removed from, the non-finalized state
    /// via the `invalidate` RPC method.
    invalidated_blocks: Arc<HashMap<block::Hash, InvalidatedBlockData>>,
```